set shortcuts for console's resize buttons

jarolrod opened the pull request:
On `master` the only way to resize the console font is to manually move your mouse and click the resize buttons. This PR introduces convenient keyboard shortcuts to resize the console font. 

The common resize shortcuts for applications are `Ctrl+=`/`Ctrl++` and `Ctrl+-`/`Ctrl+_`. This means that the resize QPushButtons need two shortcuts each, but you cannot assign multiple shortcuts to a QPushButton. See: https://doc.qt.io/qt-5/qabstractbutton.html#shortcut-prop

To get around this, we introduce a new function in `guiutil`, which connects a supplied `QKeySequence` shortcut to a `QAbstractButton`. This function can be reused in other situations where more than one shortcut is needed for a button.

| PR on macOS      | PR on Linux |
| ---------------- | ------------ |
|  ![mac-resize-shortcuts](https://user-images.githubusercontent.com/23396902/114750132-a2752580-9d21-11eb-9542-15716f2c257d.gif) | ![linux-resize-shortcuts](https://user-images.githubusercontent.com/23396902/114750165-aacd6080-9d21-11eb-8abc-5388690dcf0b.gif) |

hebasto commented:
Concept ACK.

jarolrod commented:
Updated 0caccb5 -> 3282d36

Changes:
- updated spacing to adhere to clang-format

hebasto commented:
> The common resize shortcuts for applications are `Ctrl+=`/`Ctrl++` and `Ctrl+-`/`Ctrl+_`.

Are they and `QKeySequence::ZoomIn`, `QKeySequence::ZoomOut` (https://doc.qt.io/qt-5/qkeysequence.html#StandardKey-enum) the same?

jarolrod commented:
@hebasto See: https://doc.qt.io/archives/qt-4.8/qkeysequence.html. I don't think it includes Ctrl+= or Ctrl+_

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -119,6 +120,14 @@ void setupAddressWidget(QValidatedLineEdit *widget, QWidget *parent)
+{
+    QAction* action = new QAction(parent);
+    action->setShortcuts(shortcuts);
+    parent->addAction(action);
+    parent->connect(action, &QAction::triggered, [button]() { button->animateClick(); });
```
Why the `action` is added to the `parent`, not to the `button`?

jarolrod replied:
right, will fix

jarolrod replied:
addressed in  ab95506

hebasto commented:
> @hebasto See: https://doc.qt.io/qt-5/qkeysequence.html#standard-shortcuts. I don't think it includes Ctrl+= or Ctrl+_

[Keyboard Layout Issues](https://doc.qt.io/qt-5/qkeysequence.html#keyboard-layout-issues) describe exactly this PR :)

hebasto commented:
@jarolrod 

Do you mind looking into https://github.com/hebasto/gui/commits/pr281-alt?
It seems more general and concise, no?

promag reviewed (commented):
Any reason to not use https://doc.qt.io/qt-5/qabstractbutton.html#shortcut-prop? And why we want multiple shortcuts for the same button?

hebasto commented:
> And why we want multiple shortcuts for the same button?

https://doc.qt.io/qt-5/qkeysequence.html#keyboard-layout-issues

Ctrl+Minus is, actually, Ctrl+Underscore on some keyboard layouts, no?

jarolrod commented:
@promag you can only assign one shortcut directly to a button. The reason to have multiple shortcuts is to allow for the following shortcuts on the resize buttons:
- Smaller = `Ctrl+-`, `Ctrl+_`
- Bigger = `Ctrl+=`, `Ctrl++`
These two shortcuts are common and standard for resizing and having both of them helps get around any differences in keyboard layouts.

@hebasto's approach seems to be a more focused approach as we're unlikely to need 3 shortcuts for a button. Still looking.

hebasto commented:
> ... we're unlikely to need 3 shortcuts for a button.

It will work for any number of shortcuts. But one of them could be considered as "main" one.

hebasto commented:
> Any reason to not use https://doc.qt.io/qt-5/qabstractbutton.html#shortcut-prop?

https://github.com/hebasto/gui/commits/pr281-alt is using it for a "main" shortcut.

Talkless commented on src/qt/guiutil.cpp:
```diff
@@ -119,6 +120,14 @@ void setupAddressWidget(QValidatedLineEdit *widget, QWidget *parent)
     widget->setCheckValidator(new BitcoinAddressCheckValidator(parent));
 }
 
+void setButtonShortcuts(QPushButton* button, const QList<QKeySequence>& shortcuts)
```
This function could be more generic if it would accept `QAbstractButton*` instead.

Talkless reviewed (changes requested):
Concept ACL

jarolrod commented:
Updated from ab95506 -> 2d2632a

Main Change: adopt Hebasto's [approach](https://github.com/hebasto/gui/commits/pr281-alt) as it is more focused

Changes from Hebasto's branch:
- Rename shortcut function from `AddShortcut` to `AddButtonShortcut`
- Add Doxygen comment for `AddButtonShortcut` function
- Use translator comments instead of disambiguation strings

hebasto commented on src/qt/rpcconsole.cpp:
```diff
@@ -486,8 +486,18 @@ RPCConsole::RPCConsole(interfaces::Node& node, const PlatformStyle *_platformSty
     ui->fontBiggerButton->setIcon(platformStyle->SingleColorIcon(":/icons/fontbigger"));
+    //: Main shortcut to increase the RPC console font size.
+    ui->fontBiggerButton->setShortcut(tr("Ctrl++"));
+    //: Secondary shortcut to increase the RPC console font size.
```
Maybe add a note that a secondary shortcut differs from a main one only by presence/absence of the \<Shift> key in the key sequence?

jarolrod replied:
I think this is ok as is. If we include that i'd be worried if someone translates into `Ctrl+Shift+_`.

hebasto reviewed (approved):
ACK 2d2632ac7aa0c4ebcffb5fd9d62ff9ef527fc1e6

jarolrod commented on src/qt/guiutil.h:
```diff
@@ -60,6 +62,12 @@ namespace GUIUtil
+         @param[in] button    QAbstractButton to assign shortcut to
+         @param[in] shortcut  QKeySequence to use as shortcut
+     */
+    void AddButtonShortcut(QAbstractButton* button, const QKeySequence& shortcut);
```
added an additional note in a2e122f

jarolrod commented:
updated from 2d2632a -> 2a45134

Adresses @hebasto's [comment](https://github.com/bitcoin-core/gui/pull/281#discussion_r632930904):
- Added a note to make it clear that the reason the `AddButtonShort` function exists is to get around the one shortcut at a time limitation of the button's shortcut property
- While here, adjust the formatting of my added Doxygen comment to ascribe to the formatting specified in the dev notes

Doxygen Comment:
![Screen Shot 2021-05-17 at 1 56 41 PM](https://user-images.githubusercontent.com/23396902/118546638-39b90880-b726-11eb-911e-24b61e45c5e5.png)

hebasto reviewed (approved):
re-ACK 2a45134b5694c12546d77cdff541612881f7e3e7

hebasto commented on src/qt/rpcconsole.cpp:
```diff
@@ -816,20 +826,23 @@ void RPCConsole::clear(bool clearHistory)
+                    .arg("<b>" + ui->clearButton->shortcut().toString(QKeySequence::NativeText) + "</b>") +
+                "<br>" +
+                tr("Use %1 and %2 to increase or decrease the font size.")
+                    .arg("<b>" + ui->fontBiggerButton->shortcut().toString(QKeySequence::NativeText) + "</b>")
```
But isn't the current code more readable?

Talkless replied:
Yes, maybe, though it could be something like
```
                    .arg(
                    "<b>" + ui->fontBiggerButton->shortcut().toString(QKeySequence::NativeText) + "</b>",
                    "<b>" + ui->fontSmallerButton->shortcut().toString(QKeySequence::NativeText) + "</b>"
                    ) +
```
Not _that_ bad.

jarolrod replied:
I think that should be debated here: https://github.com/bitcoin-core/gui/pull/331

hebasto replied:
@jarolrod 
> I think that should be debated here: #331

[Done](https://github.com/bitcoin-core/gui/pull/331#issuecomment-844404304) :)

Talkless replied:
Yes that ends debate.

Talkless reviewed (approved):
tACK 2a45134b5694c12546d77cdff541612881f7e3e7, tested on Debian Sid with Qt 5.15.2, shortcuts still work.
